Core/src/org/sleuthkit/autopsy/discovery/search/DomainSearchCacheLoader.java

```diff
@@ -0,0 +1,282 @@
+        final DomainCallback domainCallback = new DomainCallback(caseDb);
+        dbManager.select(domainsQuery, domainCallback);
+        
+        if (domainCallback.getSQLException() != null) {
+            throw domainCallback.getSQLException();
+        }
+        
+        if (domainCallback.getTskCoreException() != null) {
+            throw domainCallback.getTskCoreException();
+        }
+
+        return domainCallback.getResultDomains();
+    }
+
+    /**
+     * A utility method to transform filters into the necessary SQL statements
+     * for the domainsTable query. The complexity of that query requires this
+     * transformation process to be conditional. The date time filter is a good
+     * example of the type of conditional handling that follows in the method
+     * below. If no dateTime filter is supplied, then in order for the query to
+     * be correct, an additional clause needs to be added in.
+     *
+     * @return The whereClause and havingClause as a pair. These methods are one
+     * to stress that these clauses are tightly coupled.
+     */
+    Pair<String, String> createWhereAndHavingClause(List<AbstractFilter> filters) {        
+        final StringJoiner whereClause = new StringJoiner(" OR ");
+        final StringJoiner havingClause = new StringJoiner(" AND "); 
+        
+        String artifactTypeFilter = null;
+        boolean hasDateTimeFilter = false;
+        
+        for (AbstractFilter filter : filters) {       
+            if (filter instanceof ArtifactTypeFilter) {
+                artifactTypeFilter = filter.getWhereClause();
+            } else if (!(filter instanceof DataSourceFilter) && !filter.useAlternateFilter()) {
+                if (filter instanceof ArtifactDateRangeFilter) {
+                    hasDateTimeFilter = true;
+                }
+                
+                whereClause.add("(" + filter.getWhereClause() + ")");
+                havingClause.add("SUM(CASE WHEN " + filter.getWhereClause() + " THEN 1 ELSE 0 END) > 0");
+            }
+        }
+        
+        if (!hasDateTimeFilter) {
+            whereClause.add(ArtifactDateRangeFilter.createAttributeTypeClause());
+        }
+        
+        String domainAttributeFilter = "attribute_type_id = " + TSK_DOMAIN.getTypeID() +
+                " AND value_text <> ''";
+        
+        whereClause.add("(" + domainAttributeFilter + ")");
+        havingClause.add("SUM(CASE WHEN " + domainAttributeFilter + " THEN 1 ELSE 0 END) > 0");
+        
+        return Pair.of(
+                whereClause.toString() + ((artifactTypeFilter != null) ? " AND (" + artifactTypeFilter + ")" : ""),
+                havingClause.toString()
+        );
+    }
+    
+    /**
+     * Callback to handle the result set of the domain query. This callback
+     * is responsible for mapping result set rows into ResultDomain objects
+     * for display.
+     */
+    private class DomainCallback implements CaseDbAccessQueryCallback {
```

Comment:
access level of constructor and some methods in this class do not match, it looks like they should all be private I think